### ISSUE TITLE
[question] Example code?

### ISSUE THREAD

**Issue by sunny-g** [aleclarson/meteor-client#1]
Just saw you mention this and other Meteor packages [here](https://github.com/meteor/meteor/issues/3728) and was wondering if you had example code for how to use these packages, set up subscriptions, observer changes, handle accounts, etc...

I take it that this approach requires a NPM-based front-end built separate form the Meteor backend and that we'd manually require every other Meteor package we want?

**Comment by aleclarson** [comment]
> ...and was wondering if you had example code for how to use these packages, set up subscriptions, observer changes, handle accounts, etc...

Unfortunately, I don't have the time to flesh out these packages with examples and documentation. This means these packages are more of a starting point for adventurous developers to play with Meteor in a form that is compatible with NodeJS and React Native.

Of course, the end goal here is to make this accessible to all developers. It would also be nice to get the Meteor team involved, but they don't seem interested. One could assume that they are working on their own solution as we speak, but that's not absolute.

If you feel comfortable with experimenting on your own, try putting together a proof-of-concept with the packages I've provided. It would be even cooler if you could put together an example for others like you! If you run into any problems, please report them in the issues and I will help you as long as it's in my best interests.

> I take it that this approach requires a NPM-based front-end built separate form the Meteor backend and that we'd manually require every other Meteor package we want?

Correct. The packages need to be compatible with React Native's bundler, so we unfortunately cannot use Meteor's Isobuild. Though, it would be cool if we could eventually merge React Native's bundler into Isobuild as a plugin. But that's not coming anytime soon, I think.

Cheers :beers:

**Comment by sunny-g** [comment]
> Unfortunately, I don't have the time to flesh out these packages with examples and documentation. This means these packages are more of a starting point for adventurous developers to play with Meteor in a form that is compatible with NodeJS and React Native.

That's totally fair. I'm currently leaning on the [Differential meteor React-Native](http://blog.differential.com/meteor-react-native-learning-from-experience/) approach which uses hharnisc's [websocket-polyfill](https://github.com/hharnisc/react-native-meteor-websocket-polyfill), but has the downside of making you roll-your-own-everything on top of raw DDP. So I guess I have two questions:
1. What of Meteor's client-side functionality is preserved with these custom packages?
2. What changes did you make to these packages to make them work with React-Native? Is the approach amenable to someone else doing the same thing for other Meteor packages (Accounts, OAuth, etc)? I can also imagine that keeping these packages up to date with Meteor's versions alone would require a lot of maintainence.

Thanks a bunch!

**Comment by aleclarson** [comment]
> What of Meteor's client-side functionality is preserved with these custom packages?

These packages should work the exact same as their Isobuild counterparts.

> What changes did you make to these packages to make them work with React-Native?

If you look at the differences between [meteor/tracker](https://github.com/meteor/meteor/tree/devel/packages/tracker) and [aleclarson/tracker](https://github.com/aleclarson/tracker), you will see not much is different. Here is everything I typically do with each package:
- Replace the `package.js` with a `package.json`.
- Insert `require()` calls for each package dependency (Meteor's build system uses implicit dependencies, while React Native works like NodeJS and uses explicit dependencies).
- Remove any unnecessary server functionality (mostly code that uses `Meteor.isServer` or `Meteor.isClient` to execute code that only works on the server).

Additionally, I removed all tests since the changes I make are menial and I don't have the time to take out server functionality tests. But of course, it would be desirable to eventually add tests to these packages.

> I can also imagine that keeping these packages up to date with Meteor's versions alone would require a lot of maintainence.

As long as stable versions of each package exist in the form of a git tag, not much maintenance should be required. Though, it's nice to have the cutting edge (via [the `devel` branch](https://github.com/meteor/meteor/tree/devel)), but that takes more work to stay up-to-date.

To check for changes to any of these packages, I suggest using
`git diff devel upstream --name-only -- packages` in [my fork](https://github.com/aleclarson/meteor) of `meteor`.
The `devel` branch represents when I last updated these packages.
The `upstream` branch represents the newest version of `meteor`.
After finding which packages need to be updated, you still have to do `git merge` and fix merge conflicts manually.

**Comment by sunny-g** [comment]
@aleclarson that's very interesting as I wasn't expecting the changes to be so straightforward! You didn't have to remove any references to the `document` object or any browser-specific references!?

Thanks for filling me in on the changes you've made; I need to think about my specific use case (building a React-Native app alongside my Meteor server and web app while trying to share as much business logic between the two as possible) and would like to hear about the other, non-standard changes you have made, but if I end up settling with this approach, I'd love to help out in porting, updating/maintaining and testing these and other packages. 

At the very least, I can imagine a complex tool that parses the AST of the package files and programatically inserts the appropriate `require` statements and removes the `Meteor.isServer` and `Meteor.isClient` calls; converting the `packages.js` to `package.json` is something that should only have to happen once.

**Comment by aleclarson** [comment]
> You didn't have to remove any references to the document object or any browser-specific references!?

There are probably a couple changes made, but really not as many as you'd think. I'm going to create a version branch (eg: `v1.0.0`) for each of these packages. In those branches will contain the original commit that each package was forked from. This will keep the `master` branch commits in the form of `1.0.0` rather than having a massive commit history. I think this is the most organized way to do things, but maybe there's a better way you can think of. This should make it easy for you to see what changes were made to each original version stripped from [meteor/meteor](https://github.com/meteor/meteor).

> building a React-Native app alongside my Meteor server and web app while trying to share as much business logic between the two as possible

Same for me, except I'm using React Native for the web version too.

> I'd love to help out in porting, updating/maintaining and testing these and other packages.

Awesome! Pull requests are welcome. :smile:

And let me know if you decide to convert a new package so I can add it to [this list](https://github.com/aleclarson/meteor-client/wiki/Available-Packages).

> At the very least, I can imagine a complex tool that parses the AST of the package files and programatically inserts the appropriate require statements and removes the Meteor.isServer and Meteor.isClient calls; converting the packages.js to package.json is something that should only have to happen once.

That sounds great! I don't know when I'll have time to do that, though. There's definitely a lot we could do there to make this a much easier process (if not completely hands-off).

Cheers :+1:

**Comment by tgoldenberg** [comment]
@aleclarson 

Huge kudos on this project, it looks really great! Once I have my Meteor object and subscribe to a collection, how do I access the data in the collection? i.e., if I call `Meteor.subscribe('messages')`, I don't have access to the `Messages.find().fetch()` like in a Meteor app. So how then do you recommend to get that data? Thank you!!!

**Comment by aleclarson** [comment]
@tgoldenberg Use the [aleclarson/mongo](https://github.com/aleclarson/mongo) repo.

``` js
var Mongo = require('mongo');
var Messages = new Mongo.Collection('messages');
```

Cheers :grin:

**Comment by tgoldenberg** [comment]
Hey @aleclarson, thanks, that's really cool, but I mean hooking in to a publication that's in an existing Meteor app. Right now I have a Meteor app running on `localhost:3000`. I would like to point the `Meteor-client` to that separate web app and subscribe to its publications. It doesn't sound like that is what you're saying - you're saying to create a new Mongo collection on the RN client side. That makes sense, right? Right now, for the iOS app, I've been using the `ddp-client` package and establishing a connection with the localhost server and fetching data from it. Can I do the same with this project? thanks a ton!!!

**Comment by aleclarson** [comment]
@tgoldenberg Yeah, what I gave you should work. When on client-side, the `Mongo.Collection` constructor creates a cache connected to the server collection.

**Comment by tgoldenberg** [comment]
@aleclarson  All right, we're just about there :). I have the connection working (forgot to put my IP address instead of `localhost`). I have the `Meteor` server verifying that a publication is being made. However, I don't have any messages coming through, and my `meteor mongo` tells me that I definitely have 10. Sooo... here's the code I have in the RN app for Android, I'm sure we're very close. BTW, after this is worked out, I would not mind making an example app and submitting as part of a PR :)

https://gist.github.com/tgoldenberg/f6db2ca9a461773625d2

**Comment by tgoldenberg** [comment]
Also, a screenshot of the react native debugger logs - this shows that the app is connected, and then still does not register any of the subscribed messages - 
<img width="651" alt="screen shot 2015-10-27 at 3 55 20 pm" src="https://cloud.githubusercontent.com/assets/10930134/10770734/3bd06ec2-7cc3-11e5-818b-b36bb318e591.png">

**Comment by tgoldenberg** [comment]
Wow, I got it! But not by the expected way - I have `Meteor.connection._updatesForUnknownStores.messages` - but it works! Unless I can get another way that works, I plan on making a good example for this.
